File: string/palindrome-number.js
// https://leetcode.com/problems/palindrome-number/
// tags - number, string
/**
 * @param {number} x
 * @return {boolean}
 */
var isPalindrome = function(x) {
    //     const str = String(x);
    //     for (let i = 0, j = str.length-1; i < j; i++, j--) {
    //         if (str[i] !== str[j]) {
    //             return false;
    //         }
    //     }
        
    //     return true;
        
        // without converting to a string - extract all the digits into an array
        if (x < 0) {
            return false;
        }
        
        const digits = [];
        let temp = x;
        while (x > 0) {
            digits.push(x%10); // will be in reversed order - doesn't matter
            x = Math.floor(x/10);
        }
        
        for (let i = 0, j = digits.length-1; i < j; i++, j--) {
            if (digits[i] !== digits[j]) {
                return false;
            }
        }
        
        return true;
        
    };

File: string/palindrome-number.py
# https://leetcode.com/problems/palindrome-number/
# tags - string, number
class Solution:
    def isPalindrome(self, x: int) -> bool:
        # num_str = str(x)
        # for left, right in zip(num_str, reversed(num_str)):
        #     if left != right:
        #         return False
        # return True
        
        # num_str = str(x)
        # for i in range(len(num_str) // 2):
        #     if num_str[i] != num_str[-i-1]:
        #         return False
        # return True
        
        if x < 0:
            return False
        
        # Extract the digits.
        # They'll be reversed but that doesn't matter.
        digits = []
        while x > 0:
            digits.append(x % 10)
            x //= 10
        
        # for i in range(len(digits) // 2):
        #     if digits[i] != digits[-i-1]:
        #         return False
        # return True
        
        i, j = 0, len(digits)-1
        while i < j:
            if digits[i] != digits[j]:
                return False
            i += 1
            j -= 1
        return True
        
        